lexxpluss_apps/src/bmu_controller.cpp

```diff
@@ -0,0 +1,209 @@
+                    msg.bmu_alarm1, msg.bmu_alarm2,
+                    msg.max_cell_voltage.value, msg.max_cell_voltage.id, msg.min_cell_voltage.value, msg.min_cell_voltage.id,
+                    msg.manufacturing, msg.inspection, msg.serial);
+    }
+
+private:
+    bool handler_bmu(zcan_frame &frame) {
+        bool result{false};
+        if (frame.id == 0x100) {
+            msg.mod_status1 = frame.data[0];
+            msg.bmu_status = frame.data[1];
+            msg.asoc = frame.data[2];
+            msg.rsoc = frame.data[3];
+            msg.soh = frame.data[4];
+            msg.fet_temp = (frame.data[5] << 8) | frame.data[6];
+        } else if (frame.id == 0x101) {
+            msg.pack_current = (frame.data[0] << 8) | frame.data[1];
+            msg.charging_current = (frame.data[2] << 8) | frame.data[3];
+            msg.pack_voltage = (frame.data[4] << 8) | frame.data[5];
+            msg.mod_status2 = frame.data[6];
+        } else if (frame.id == 0x103) {
+            msg.design_capacity = (frame.data[0] << 8) | frame.data[1];
+            msg.full_charge_capacity = (frame.data[2] << 8) | frame.data[3];
+            msg.remain_capacity = (frame.data[4] << 8) | frame.data[5];
+        } else if (frame.id == 0x110) {
+            msg.max_voltage.value = (frame.data[0] << 8) | frame.data[1];
+            msg.max_voltage.id = frame.data[2];
+            msg.min_voltage.value = (frame.data[4] << 8) | frame.data[5];
+            msg.min_voltage.id = frame.data[6];
+        } else if (frame.id == 0x111) {
+            msg.max_temp.value = (frame.data[0] << 8) | frame.data[1];
+            msg.max_temp.id = frame.data[2];
+            msg.min_temp.value = (frame.data[4] << 8) | frame.data[5];
+            msg.min_temp.id = frame.data[6];
+        } else if (frame.id == 0x112) {
+            msg.max_current.value = (frame.data[0] << 8) | frame.data[1];
+            msg.max_current.id = frame.data[2];
+            msg.min_current.value = (frame.data[4] << 8) | frame.data[5];
+            msg.min_current.id = frame.data[6];
+        } else if (frame.id == 0x113) {
+            msg.bmu_fw_ver = frame.data[0];
+            msg.mod_fw_ver = frame.data[1];
+            msg.serial_config = frame.data[2];
+            msg.parallel_config = frame.data[3];
+            msg.bmu_alarm1 = frame.data[4];
+            msg.bmu_alarm2 = frame.data[5];
+        } else if (frame.id == 0x120) {
+            msg.min_cell_voltage.value = (frame.data[0] << 8) | frame.data[1];
+            msg.min_cell_voltage.id = frame.data[2];
+            msg.max_cell_voltage.value = (frame.data[4] << 8) | frame.data[5];
+            msg.max_cell_voltage.id = frame.data[6];
+        } else if (frame.id == 0x130) {
+            msg.manufacturing = (frame.data[0] << 8) | frame.data[1];
+            msg.inspection = (frame.data[2] << 8) | frame.data[3];
+            msg.serial = (frame.data[4] << 8) | frame.data[5];
+            result = true;
+        }
```

Comment:
Is it safe to expect to receive 0x130 last?
Even if it is OK to receive 0x130 last, is there any concern that the process will start with receiving another CANID instead of 0x100 for the first time? ( There is a concern of getting weird data the first time. )

0x130を最後に受信することを期待しても大丈そうな仕様でしょうか?
もし、最後に受信で良い場合も、初回は0x100ではなくほかのCANIDの受信から処理が始まる懸念はないでしょうか? ( 初回はヘンテコなデータを取得する懸念がある。 )

Reply:
新ファイルですが、西東さんの実装のままです。というだけの理由ですが実績のあるコードなので一旦このままとしたいです。
データシートを見る限りはまとまってやってくるので、最初の１回以外は問題なさそう。
ただ、なんらかコーナーケース発生時にバグを生みやすい可能性はあるので、改善できないか考えてみます。
また西東さんにも意見を聞いてみます。